controller/handler_ctrl/bind.go

```diff
@@ -99,3 +119,60 @@ func (self *ctrlChannelLatencyHandler) HandleLatency(latencyType latency.Type, e
 		self.queueTimeHistogram.Update(elapsed.Nanoseconds())
 	}
 }
+
+type heartbeatCallback struct {
+	latencyMetric    metrics.Histogram
+	queueTimeMetric  metrics.Histogram
+	firstSent        int64
+	lastResponse     int64
+	ch               channel.Channel
+	latencySemaphore concurrenz.Semaphore
+}
+
+func (self *heartbeatCallback) HeartbeatTx(int64) {
+	if self.firstSent == 0 {
+		self.firstSent = time.Now().UnixMilli()
+	}
+}
+
+func (self *heartbeatCallback) HeartbeatRx(int64) {}
+
+func (self *heartbeatCallback) HeartbeatRespTx(int64) {}
+
+func (self *heartbeatCallback) HeartbeatRespRx(ts int64) {
+	now := time.Now()
+	self.lastResponse = now.UnixNano()
```

Comment:
Sorry, I wasn't clear, I meant the time.Now().UnixNano() below should just be `now.UnixNano()`. The `now.UnixMillis()` should stay as it was

Reply:
Oops, either would have made sense to me so didn't think much about it. Changed.